src/client_impl_test.cc

```diff
@@ -0,0 +1,319 @@
+
+  ~MockQuotaTransport() {}
+
+  // The done callback is called right away (in place).
+  void QuotaWithInplaceCallback(const QuotaRequest &request,
+                                QuotaResponse *response, DoneFunc on_done) {
+    quota_request_ = request;
+    if (quota_response_) {
+      *response = *quota_response_;
+    }
+    on_done(done_status_);
+  }
+
+  // Saved report_request from mocked Transport::Report() call.
+  QuotaRequest quota_request_;
+  // If not NULL, the report response to send for mocked Transport::Report()
+  // call.
+  QuotaResponse *quota_response_;
+
+  // The status to send in on_done call back for Check() or Report().
+  Status done_status_;
+};
+}  // namespace
+
+class MixerClientImplTest : public ::testing::Test {
+ public:
+  void SetUp() {
+    ASSERT_TRUE(TextFormat::ParseFromString(kCheckRequest1, &check_request1_));
+    ASSERT_TRUE(TextFormat::ParseFromString(kSuccessCheckResponse1,
+                                            &pass_check_response1_));
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kReportRequest1, &report_request1_));
+
+    ASSERT_TRUE(TextFormat::ParseFromString(kQuotaRequest1, &quota_request1_));
+
+    MixerClientOptions options(
+        CheckOptions(1 /*entries */, 500 /* refresh_interval_ms */,
+                     1000 /* expiration_ms */),
+        ReportOptions(1 /* entries */, 500 /*flush_interval_ms*/),
+        QuotaOptions(1 /* entries */, 500 /*flush_interval_ms*/,
+                     1000 /* expiration_ms */));
+    options.check_transport = mock_check_transport_.GetFunc();
+    options.report_transport = mock_report_transport_.GetFunc();
+    options.quota_transport = mock_quota_transport_.GetFunc();
+    client_ = CreateMixerClient(options);
+  }
+
+  CheckRequest check_request1_;
+  CheckResponse pass_check_response1_;
+  CheckResponse error_check_response1_;
+
+  ReportRequest report_request1_;
+  QuotaRequest quota_request1_;
+
+  MockCheckTransport mock_check_transport_;
+  MockReportTransport mock_report_transport_;
+  MockQuotaTransport mock_quota_transport_;
+
+  std::unique_ptr<MixerClient> client_;
+};
+
+TEST_F(MixerClientImplTest, TestNonCachedCheckWithInplaceCallback) {
+  // Calls a Client::Check, the request is not in the cache
+  // Transport::Check() is called.  It will send a successful check response
+  EXPECT_CALL(mock_check_transport_, Check(_, _, _))
+      .WillOnce(Invoke(&mock_check_transport_,
+                       &MockCheckTransport::CheckWithInplaceCallback));
+
+  mock_check_transport_.done_status_ = Status::OK;
+
+  CheckResponse check_response;
+  Status done_status = Status::UNKNOWN;
+  client_->Check(check_request1_, &check_response,
+                 [&done_status](Status status) { done_status = status; });
+
+  // Since it is not cached, transport should be called.
+  EXPECT_TRUE(MessageDifferencer::Equals(mock_check_transport_.check_request_,
+                                         check_request1_));
+}
```

Comment:
You can also assert done_status == mock_check_transport_.done_status_

for all 3 calls.